### PR TITLE
Add PipeOp Tags

### PR DESCRIPTION
**Pull request by pfistfl**
Very rudimentary implementation, but workable.
This would add the following tags:
**Updated**
```
> as.data.table(mlr_pipeops)[,c(1,3)]
                key                             tags
 1:          boxcox                   data transform
 2:          branch                             meta
 3:           chunk                             meta
 4:  classbalancing   imbalanced data,data transform
 5:      classifavg                         ensemble
 6:    classweights   imbalanced data,data transform
 7:        colapply                   data transform
 8: collapsefactors                   data transform
 9:            copy                             meta
10:          encode            encode,data transform
11:    encodeimpact            encode,data transform
12:      encodelmer            encode,data transform
13:    featureunion                         ensemble
14:          filter feature selection,data transform
15:      fixfactors         robustify,data transform
16:         histbin                   data transform
17:             ica                   data transform
18:      imputehist                         missings
19:      imputemean                         missings
20:    imputemedian                         missings
21:    imputenewlvl                         missings
22:    imputesample                         missings
23:       kernelpca                   data transform
24:         learner                          learner
25:      learner_cv          ensemble,data transform
26:         missind          missings,data transform
27:     modelmatrix                   data transform
28:          mutate                   data transform
29:             nop                             meta
30:             pca                   data transform
31:     quantilebin                   data transform
32:         regravg                         ensemble
33: removeconstants         robustify,data transform
34:           scale                   data transform
35:     scalemaxabs                   data transform
36:      scalerange                   data transform
37:          select feature selection,data transform
38:           smote   imbalanced data,data transform
39:     spatialsign                   data transform
40:       subsample                   data transform
41:        unbranch                             meta
42:      yeojohnson                   data transform

```

### CODE REVIEW

**Comment by pfistfl** [comment]
Perhaps all "impute" po's should also have `data transform`

**Comment by pfistfl** [comment]
` classweights` is currently a little inconclusive, also add it to `imbalanced`?

File: R/typecheck.R [hunk 1]
```diff
@@ -240,3 +240,18 @@ get_autoconverter = function(target) {
 
   NULL
 }
+
+# Check that a vector of tags are valid tags for `PipeOp's`.
+# @param tags [`character`]: A list of tags.
+# @return a character vector of tasks
+assert_tag = function(tags) {
+  tags = assert_character(tolower(tags))
+  valid_tags = c("abstract", "meta", "missings", "feature selection", "imbalanced data",
```
Comment (mb706) [inline R/typecheck.R]: put this in mlr3 reflections

File: R/PipeOp.R [hunk 2]
```diff
@@ -206,6 +207,7 @@ PipeOp = R6Class("PipeOp",
       self$input = assert_connection_table(input)
       self$output = assert_connection_table(output)
       self$packages = assert_character(packages, any.missing = FALSE, unique = TRUE)
+      self$tags = "abstract"
```
Comment (mb706) [inline R/PipeOp.R]: empty tag for base pipeop

File: R/PipeOpBoxCox.R [hunk 3]
```diff
@@ -72,6 +72,7 @@ PipeOpBoxCox = R6Class("PipeOpBoxCox",
       ))
       super$initialize(id, param_set = ps, param_vals = param_vals,
         packages = "bestNormalize")
+      private$add_tags("feature type: numeric")
```
Comment (mb706) [inline R/PipeOpBoxCox.R]: give this as constructor argument

File: R/zzz.R [hunk 4]
```diff
@@ -9,9 +9,22 @@
...
 
 .onLoad = function(libname, pkgname) {  # nocov start
-
+  register_mlr3()
+  setHook(packageEvent("mlr3", "onLoad"), function(...) register_mlr3(), action = "append")
   backports::import(pkgname)
-
 }  # nocov end
+
+.onUnload = function(libpath) { # nocov start
```
Comment (mb706) [inline R/zzz.R]: what is this for?

File: R/PipeOp.R [hunk 5]
```diff
@@ -191,8 +191,9 @@ PipeOp = R6Class("PipeOp",
     input = NULL,
     output = NULL,
     .result = NULL,
+    tags = NULL,
 
-    initialize = function(id, param_set = ParamSet$new(), param_vals = list(), input, output, packages = character(0)) {
+    initialize = function(id, param_set = ParamSet$new(), param_vals = list(), input, output, packages = character(0), tags = character()) {
```
Comment (mb706) [inline R/PipeOp.R]: can we use `character(0)` for consistency?

File: R/typecheck.R [hunk 6]
```diff
@@ -240,3 +240,14 @@ get_autoconverter = function(target) {
...
   NULL
 }
+
+# Check that a vector of tags are valid tags for `PipeOp's`.
+# See `mlr_reflections$pipeops$valid_tags` for a list of valid tags.
+# @param tags [`character`]: A list of tags.
+# @return a character vector of tasks
+assert_tag = function(tags) {
+  if (length(tags) == 0) tags = "abstract" else tags = assert_character(tolower(tags))
+  valid_tags = c(mlr_reflections$pipeops$valid_tags, paste0("feature type: ", mlr_reflections$task_feature_types))
```
Comment (mb706) [inline R/typecheck.R]: I assume the `feature type:` part is obsolete now?

File: R/typecheck.R [hunk 7]
```diff
@@ -240,3 +240,14 @@ get_autoconverter = function(target) {
 
   NULL
 }
+
+# Check that a vector of tags are valid tags for `PipeOp's`.
+# See `mlr_reflections$pipeops$valid_tags` for a list of valid tags.
+# @param tags [`character`]: A list of tags.
+# @return a character vector of tasks
+assert_tag = function(tags) {
```
Comment (mb706) [inline R/typecheck.R]: I don't know if we really need this kind of function; it is called only once, and mlr3::Learner just uses `assert_subset(..., mlr_reflections$...)` in its place.

File: R/zzz.R (hunk 4 above)
Comment (pfistfl) [reply R/zzz.R]: I think this makes sure that unloading the package cleanly removes the additions made by `register_mlr3`.
I copied this over from https://github.com/mlr-org/mlr3proba/blob/master/R/zzz.R

**Comment by pfistfl** [comment]
tags and feature types should be added for all `TaskPreproc` pipelines. 
Others often do not have any feature types.

File: R/PipeOpTaskPreproc.R [hunk 8]
```diff
@@ -152,7 +152,8 @@ PipeOpTaskPreproc = R6Class("PipeOpTaskPreproc",
   inherit = PipeOp,
 
   public = list(
-    initialize = function(id, param_set = ParamSet$new(), param_vals = list(), can_subset_cols = TRUE, packages = character(0), task_type = "Task") {
+    initialize = function(id, param_set = ParamSet$new(), param_vals = list(), can_subset_cols = TRUE,
+      packages = character(0), task_type = "Task", tags = NULL, feature_types = mlr_reflections$task_feature_types) {
```
Comment (mb706) [inline R/PipeOpTaskPreproc.R]: could you add documentation to the feature_types parameter in the PipeOpTaskPreproc$new part?

File: R/PipeOpTaskPreproc.R [hunk 9]
```diff
@@ -162,10 +163,11 @@ PipeOpTaskPreproc = R6Class("PipeOpTaskPreproc",
           param_set = c(param_set, alist(private$.affectcols_ps))
         }
       }
+      private$.feature_types = feature_types
```
Comment (mb706) [inline R/PipeOpTaskPreproc.R]: assert_subset here

File: R/PipeOpTaskPreproc.R [hunk 10]
```diff
@@ -247,10 +249,21 @@ PipeOpTaskPreproc = R6Class("PipeOpTaskPreproc",
...
 
-    select_cols = function(task) task$feature_names
+    select_cols = function(task) {
+      task$feature_types[get("type") %in% self$feature_types, get("id")]
+    }
+  ),
+  active = list(
+    feature_types = function(types) {
+      if (!missing(types)) {
+        private$.feature_types = assert_subset(types, mlr_reflections$task_feature_types)
```
Comment (mb706) [inline R/PipeOpTaskPreproc.R]: The user should not change feature_types (if he wants to restrict action, it should be with the 'affect_columns' parameter), so if (!missing) should ideally throw an error.

File: R/PipeOp.R [hunk 11]
```diff
@@ -191,8 +191,9 @@ PipeOp = R6Class("PipeOp",
     input = NULL,
     output = NULL,
     .result = NULL,
+    tags = NULL,
 
-    initialize = function(id, param_set = ParamSet$new(), param_vals = list(), input, output, packages = character(0)) {
+    initialize = function(id, param_set = ParamSet$new(), param_vals = list(), input, output, packages = character(0), tags = character(0)) {
```
Comment (mb706) [inline R/PipeOp.R]: could you add documentation to the `tags` parameter of the `new()` function of this class?

File: R/zzz.R (hunk 4 above)
Comment (mb706) [reply R/zzz.R]: I guess this is how this is done now...

**Pull request merged**